### ISSUE TITLE
Overload TypeVar Bug

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#2568]
**Describe the bug**

```python
@overload
def to_list(value: Iterable[T]) -> list[T]:
    ...


@overload
def to_list(value: T) -> list[T]:
    ...


def to_list(value: Iterable[T] | T) -> list[T]: # error
    if isinstance(value, Iterable):
        return list(value)
    else:
        return [value]


def func1(x: list[int] | int):
    y: list[int] = to_list(x) # another error
    reveal_type(y)
```

**To Reproduce**
Run pyright/pylance. Two error messages are produced. The errors are, 

Overloaded function implementation is not consistent with signature of overload 1
  Function return type "list[T@to_list]" is incompatible with type "list[Iterable[T@to_list]]"
    TypeVar "_T@list" is invariant
      Type "T@to_list" cannot be assigned to type "Iterable[T@to_list]"

Expression of type "list[list[int] | int]" cannot be assigned to declared type "list[int]"
  TypeVar "_T@list" is invariant
    Type "list[int] | int" cannot be assigned to type "int"
      "list[int]" is incompatible with "int"

**Expected behavior**
No errors like mypy.

**VS Code extension or command-line**
pyright 1.1.187.

**Additional context**
The second error message I bisected to near pyright 1.1.166. First error message seems to have always been here as testing with old versions I still see it.

**Comment by erictraut** [comment]
Looking at this more, I think the current behavior correct, or at least it is expected.

The problem is that you have created an ambiguous set of overloads.

Pyright resolves overloads by evaluating each overload in order. If the arguments for an overload matches, that is the "winner". If none of the overloads match but one or more of the arguments are union types, the unions are "split" into their constituent subtypes, and the argument lists are then independently matched against the overloads. 

In your sample, the type `list[int] | int` is first compared against `Iterable[T]` in the first overload. That's not a compatible type, so the second overload is then considered. Since `list[int] | int` is assignable to `T`, this overload is a match. It becomes the "winner", and the resulting return type is `list[list[int] | int`.

I'm curious why you thought that an overload was needed in this example. It's unnecessary because the return type is the same for both overloads. Perhaps your code was more complex than this sample demonstrates?

The correct way to annotate this sample is to simply delete the overloads. This eliminates the errors.

```python
from typing import Iterable, TypeVar

T = TypeVar("T")

def to_list(value: Iterable[T] | T) -> list[T]:
    if isinstance(value, Iterable):
        return list(value)
    else:
        return [value]

def func1(x: list[int] | int):
    y: list[int] = to_list(x)
    reveal_type(y)
```

**Comment by hmc-cs-mdrissi** [comment]
https://github.com/python/typing/discussions/943#discussioncomment-1641595 I asked on typing discussion how to type hint this function and was recommended overload approach. I then tested the overload approach works with mypy and reveal_type behaves as expected.

The other approach of not using overloads does work and was also mentioned in the discussion. If this part of the overload behavior is undefined behavior across mypy vs pyright then the typing discussion answer should be changed to not recommend overloads here. Or at least be more specific on when that's right approach as if func1 didn't have a union of arguments and just had list[int] or int I think that would be ok with overload approach. I did originally use non-overload approach (although initially was missing y: list[int] as without that clue it picks other type).

**Comment by erictraut** [comment]
As a general rule, if you find that you're creating overloads that all return the same type, you're probably reaching for the wrong tool. Overloads are meant to handle the case where return types depend on the arguments.
